ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java

```diff
@@ -221,13 +222,150 @@ public ReplicaSyncUpManager getReplicaSyncUpManager() {
+    Map<String, Map<String, String>> mountPathToDiskInfos = instanceConfig.getRecord().getMapFields();
+    Map<String, String> diskInfo = mountPathToDiskInfos.get(replicaId.getMountPath());
+    boolean newReplicaInfoAdded = false;
+    boolean duplicateFound = false;
+    if (diskInfo != null) {
+      // add replica to an existing disk (need to sort replicas by partition id)
+      String replicasStr = diskInfo.get(ClusterMapUtils.REPLICAS_STR);
+      String[] replicaInfos = replicasStr.split(ClusterMapUtils.REPLICAS_DELIM_STR);
+      StringBuilder replicasStrBuilder = new StringBuilder();
+      long idToAdd = Long.valueOf(partitionName);
+      for (String replicaInfo : replicaInfos) {
+        String[] infos = replicaInfo.split(ClusterMapUtils.REPLICAS_STR_SEPARATOR);
+        long currentId = Long.valueOf(infos[0]);
+        if (currentId == idToAdd) {
+          logger.info("Partition {} is already on instance {}, skipping adding it into InstanceConfig in Helix.",
+              partitionName, instanceName);
+          duplicateFound = true;
+          break;
+        } else if (currentId < idToAdd || newReplicaInfoAdded) {
+          replicasStrBuilder.append(replicaInfo).append(ClusterMapUtils.REPLICAS_DELIM_STR);
+        } else {
+          // newReplicaInfo already contains delimiter, no need to append REPLICAS_DELIM_STR
+          replicasStrBuilder.append(newReplicaInfo);
+          replicasStrBuilder.append(replicaInfo).append(ClusterMapUtils.REPLICAS_DELIM_STR);
+          newReplicaInfoAdded = true;
+        }
+      }
+      if (!duplicateFound && !newReplicaInfoAdded) {
+        // this means new replica id is larger than all existing replicas' ids
+        replicasStrBuilder.append(newReplicaInfo);
+        newReplicaInfoAdded = true;
+      }
+      if (newReplicaInfoAdded) {
+        diskInfo.put(ClusterMapUtils.REPLICAS_STR, replicasStrBuilder.toString());
+        mountPathToDiskInfos.put(replicaId.getMountPath(), diskInfo);
+      }
+    } else {
+      // add replica onto a brand new disk
+      Map<String, String> diskInfoToAdd = new HashMap<>();
+      diskInfoToAdd.put(ClusterMapUtils.DISK_CAPACITY_STR,
+          Long.toString(replicaId.getDiskId().getRawCapacityInBytes()));
+      diskInfoToAdd.put(ClusterMapUtils.DISK_STATE, ClusterMapUtils.AVAILABLE_STR);
+      diskInfoToAdd.put(ClusterMapUtils.REPLICAS_STR, newReplicaInfo);
+      mountPathToDiskInfos.put(replicaId.getMountPath(), diskInfoToAdd);
+      newReplicaInfoAdded = true;
+    }
+    if (newReplicaInfoAdded) {
+      // we update InstanceConfig only when new replica info is added (skip updating if replica is already present)
+      instanceConfig.getRecord().setMapFields(mountPathToDiskInfos);
+      logger.info("Updating config: {} in Helix by adding partition {}", instanceConfig, partitionName);
+      additionResult = helixAdmin.setInstanceConfig(clusterName, instanceName, instanceConfig);
+    }
+    return additionResult;
```

Comment:
looks like additionResult is true even if partition is already present in an instance. Is that expected behavior? Should we return false in this case, as really we didnt add anything here.